urllib3/connection.py

```diff
@@ -299,7 +302,15 @@ def connect(self):
                     'for details.)'.format(hostname)),
                     SubjectAltNameWarning
                 )
-            match_hostname(cert, self.assert_hostname or hostname)
+            try:
+                match_hostname(cert, self.assert_hostname or hostname)
+            except CertificateError as e:
+                log.error(
+                    'Certificate did not match expected hostname: %s. '
+                    'Certificate: %s', asserted_hostname, cert
+                )
+                e.cert = cert
+                raise e
```

Comment:
> You could investigate whether leaving the patch in and changing to a naked
> raise statement still successfully patches the cert onto the exception.
> That would be the best of both worlds I think.

Just tested and it works so I updated it. Thanks!

Note that because exceptions get wrapped by other exceptions, it's somewhat of a pain to get at that `cert` attribute, though possible is better than not possible.

```
(Pdb++) e.cert
*** AttributeError: 'SSLError' object has no attribute 'cert'
(Pdb++) e.args[0].cert
*** AttributeError: 'SSLError' object has no attribute 'cert'
(Pdb++) e.args[0].args[0].cert
{'subjectAltName': (('IP Address', '127.0.0.1'), ('IP Address', '10.10.194.36'), ('IP Address', '10.10.194.37'), ('IP Address', '10.10.194.38')), 'notBefore': u'Mar 17 01:18:27 2016 GMT', 'serialNumber': u'9BFDB58B08E7DCCA', 'notAfter': 'Mar 15 01:18:27 2026 GMT', 'version': 3L, 'subject': ((('countryName', u'US'),), (('stateOrProvinceName', u'California'),), (('localityName', u'Palo Alto'),), (('organizationName', u'DevOps'),), (('commonName', u'vault.service.trouper'),)), 'issuer': ((('countryName', u'US'),), (('stateOrProvinceName', u'California'),), (('localityName', u'Palo Alto'),), (('organizationName', u'DevOps'),), (('commonName', u'vault.service.trouper'),))}
```